python/ray/data/_internal/batcher.py

```diff
@@ -78,3 +121,203 @@ def next_batch(self) -> Block:
         # blocks consumed on the next batch extraction.
         self._buffer = leftover
```

Comment:
Good catch! I think what's needed is actually
```suggestion
        self._buffer = leftover
        self._buffer_size -= self._batch_size
```
since the buffer size is in the number of underlying rows rather than the number of blocks.

This is most likely a culprit of a failing test that I was about to debug, so an extra thank you to you. 😄